### ISSUE TITLE
Poor performance in browser

### ISSUE THREAD

**Issue by avalanche1** [justadudewhohacks/face-api.js#417]
`fullFaceDescriptions = await faceapi.detectAllFaces(input).withFaceLandmarks().withFaceDescriptors()
` takes randomly 5-15 seconds. Is it expected when doing face detection in browser?

**Comment by justadudewhohacks** [comment]
No, depending on the device you are running this on, this should only take a few miliseconds.

`faceapi.detectAllFaces(input)` by default also by default uses the SSDMobilenetV1 model, which can potentially be slower on some devices. You can try out the tiny face detector instead (see README).

Also make sure the WebGL backend is utilized: `console.log(tf.getBackend())`.

**Comment by avalanche1** [comment]
what is `tf` and how to import it?

**Comment by zjlovezj** [comment]
```js
import * as faceapi from "face-api.js";
faceapi.tf.getBackend()
```

@avalanche1

**Comment by zjlovezj** [comment]
@justadudewhohacks 
By saying "a few miliseconds", you mean "a few hundreds miliseconds"? :)
I'm using face-api.js in a non-critical page and I got some stats.
conf:
* tiny face detector
* face-api.js@0.20.1 

It's roughly like these:
Total users: about 20k.
75% took below 500ms.
10% took (500, 1500)ms.
5% took 1000ms+.
The extreme case did take more than 10 seconds.

I've tried face-api.js@0.20.1 on some computer, performance seems go down a little bit.

What's your suggestions to improve performance? or gain some other stats?

BTW, here are some cases I don't understand.
```
// took 1400ms to finish a single detection
webgl: true; tf_backend: webgl; os: win7; cpu: Intel(R) Core(TM) i5-4200U CPU @ 1.60GHz

// took 2400ms to finish a single detection
webgl: null; tf_backend: cpu; os: win10; cpu: Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz
```

Thanks

**Comment by justadudewhohacks** [comment]
@zjlovezj, the tiny face detector model on my desktop machine with webgl backend and an input size of 160 runs at about 60 fps, so the time for inference is about 20ms or less. On my android it runs at 15 fps. The inference time obviously depends on the device.

The question is, do your metrics include the shader compilation times? You should exclude the very first forward pass from your measurements.

**Comment by zjlovezj** [comment]
Yes. I did exclude the first detection. And the numbers are average of 20 times of detection.

**Comment by avalanche1** [comment]
@zjlovezj I Think a repro would help; I'll try to hack up smth as well

**Comment by zjlovezj** [comment]
@avalanche1 I don't think a repo would help much on this performance issue. The results are all different on different machine with different configuration.
